File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/Location.java
package wangdaye.com.geometricweather.common.basic.models;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.TimeZone;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;
import wangdaye.com.geometricweather.common.utils.LanguageUtils;

/**
 * Location.
 * */

public class Location
        implements Parcelable {

    private final String cityId;

    private final float latitude;
    private final float longitude;
    private final TimeZone timeZone;

    private final String country;
    private final String province;
    private final String city;
    private final String district;

    @Nullable private Weather weather;
    private final WeatherSource weatherSource;

    private final boolean currentPosition;
    private final boolean residentPosition;
    private final boolean china;

    private static final String NULL_ID = "NULL_ID";
    public static final String CURRENT_POSITION_ID = "CURRENT_POSITION";

    public Location(Location src, WeatherSource weatherSource) {
        this(src.cityId, src.latitude, src.longitude, src.timeZone, src.country, src.province,
                src.city, src.district, src.weather, weatherSource, src.currentPosition,
                src.residentPosition, src.china);
    }

    public Location(Location src, boolean currentPosition, boolean residentPosition) {
        this(src.cityId, src.latitude, src.longitude, src.timeZone, src.country, src.province,
                src.city, src.district, src.weather, src.weatherSource,
                currentPosition, residentPosition, src.china);
    }

    public Location(Location src,
                    float latitude, float longitude, TimeZone timeZone) {
        this(src.cityId, latitude, longitude, timeZone, src.country, src.province, src.city, src.district,
                src.weather, src.weatherSource, src.currentPosition, src.residentPosition, src.china);
    }

    public Location(String cityId, float latitude, float longitude, TimeZone timeZone,
                    String country, String province, String city, String district,
                    @Nullable Weather weather, WeatherSource weatherSource,
                    boolean currentPosition, boolean residentPosition, boolean china) {
        this.cityId = cityId;
        this.latitude = latitude;
        this.longitude = longitude;
        this.timeZone = timeZone;
        this.country = country;
        this.province = province;
        this.city = city;
        this.district = district;
        this.weather = weather;
        this.weatherSource = weatherSource;
        this.currentPosition = currentPosition;
        this.residentPosition = residentPosition;
        this.china = china;
    }

    public static Location buildLocal() {
        return new Location(
                NULL_ID,
                0, 0, TimeZone.getDefault(),
                "", "", "", "",
                null, WeatherSource.ACCU,
                true, false, false
        );
    }

    public static Location buildDefaultLocation() {
        return new Location(
                "101924",
                39.904000f, 116.391000f, TimeZone.getTimeZone("Asia/Shanghai"),
                "中国", "直辖市", "北京", "",
                null, WeatherSource.ACCU,
                false, false, true
        );
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        if (!getFormattedId().equals(
                ((Location) o).getFormattedId()
        )) {
            return false;
        }

        Weather thisWeather = getWeather();
        Weather otherWeather = ((Location) o).getWeather();
        if (thisWeather == null && otherWeather == null) {
            return true;
        }
        if (thisWeather != null && otherWeather != null) {
            return thisWeather.getBase().getTimeStamp()
                    == otherWeather.getBase().getTimeStamp();
        }
        return false;
    }

    @Override
    public int hashCode() {
        return getFormattedId().hashCode();
    }

    public String getFormattedId() {
        return isCurrentPosition()
                ? CURRENT_POSITION_ID
                : (cityId + "&" + weatherSource.name());
    }

    public boolean isCurrentPosition() {
        return currentPosition;
    }

    public boolean isResidentPosition() {
        return residentPosition;
    }

    public boolean isUsable() {
        return !cityId.equals(NULL_ID);
    }

    public boolean canUseChineseSource() {
        return LanguageUtils.isChinese(city) && china;
    }

    public String getCityId() {
        return cityId;
    }

    public float getLatitude() {
        return latitude;
    }

    public float getLongitude() {
        return longitude;
    }

    public TimeZone getTimeZone() {
        return timeZone;
    }

    public String getCountry() {
        return country;
    }

    public String getProvince() {
        return province;
    }

    public String getCity() {
        return city;
    }

    public String getDistrict() {
        return district;
    }

    public String getCityName(Context context) {
        if (!TextUtils.isEmpty(district) && !district.equals("市辖区") && !district.equals("无")) {
            return district;
        } else if (!TextUtils.isEmpty(city) && !city.equals("市辖区")) {
            return city;
        } else if (!TextUtils.isEmpty(province)) {
            return province;
        } else if (currentPosition) {
            return context.getString(R.string.current_location);
        } else {
            return "";
        }
    }

    @NonNull
    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder(getCountry() + " " + getProvince());
        if (!getProvince().equals(getCity())
                && !TextUtils.isEmpty(getCity())) {
            builder.append(" ").append(getCity());
        }
        if (!getCity().equals(getDistrict())
                && !TextUtils.isEmpty(getDistrict())) {
            builder.append(" ").append(getDistrict());
        }
        return builder.toString();
    }

    public boolean hasGeocodeInformation() {
        return !TextUtils.isEmpty(country)
                || !TextUtils.isEmpty(province)
                || !TextUtils.isEmpty(city)
                || !TextUtils.isEmpty(district);
    }

    @Nullable
    public Weather getWeather() {
        return weather;
    }

    public void setWeather(@Nullable Weather weather) {
        this.weather = weather;
    }

    public WeatherSource getWeatherSource() {
        return weatherSource;
    }

    public boolean isChina() {
        return china;
    }

    private static boolean isEquals(@Nullable String a, @Nullable String b) {
        if (TextUtils.isEmpty(a) && TextUtils.isEmpty(b)) {
            return true;
        } else if (!TextUtils.isEmpty(a) && !TextUtils.isEmpty(b)) {
            return a.equals(b);
        } else {
            return false;
        }
    }

    public static List<Location> excludeInvalidResidentLocation(Context context, List<Location> list) {
        Location currentLocation = null;
        for (Location l : list) {
            if (l.isCurrentPosition()) {
                currentLocation = l;
                break;
            }
        }

        List<Location> result = new ArrayList<>(list.size());
        if (currentLocation == null) {
            result.addAll(list);
        } else {
            for (Location l : list) {
                if (!l.isResidentPosition() || !l.isCloseTo(context, currentLocation)) {
                    result.add(l);
                }
            }
        }
        return result;
    }

    private boolean isCloseTo(Context c, Location location) {
        if (cityId.equals(location.getCityId())) {
            return true;
        }
        if (isEquals(province, location.province)
                && isEquals(city, location.city)) {
            return true;
        }
        if (isEquals(province, location.province)
                && getCityName(c).equals(location.getCityName(c))) {
            return true;
        }
        return Math.abs(latitude - location.latitude) < 0.8
                && Math.abs(longitude - location.longitude) < 0.8;
    }

    public boolean isDaylight() {
        if (weather != null ) {
            return weather.isDaylight(getTimeZone());
        }

        return DisplayUtils.isDaylight(getTimeZone());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.cityId);
        dest.writeFloat(this.latitude);
        dest.writeFloat(this.longitude);
        dest.writeSerializable(this.timeZone);
        dest.writeString(this.country);
        dest.writeString(this.province);
        dest.writeString(this.city);
        dest.writeString(this.district);
        dest.writeInt(this.weatherSource == null ? -1 : this.weatherSource.ordinal());
        dest.writeByte(this.currentPosition ? (byte) 1 : (byte) 0);
        dest.writeByte(this.residentPosition ? (byte) 1 : (byte) 0);
        dest.writeByte(this.china ? (byte) 1 : (byte) 0);
    }

    protected Location(Parcel in) {
        this.cityId = in.readString();
        this.latitude = in.readFloat();
        this.longitude = in.readFloat();
        this.timeZone = (TimeZone) in.readSerializable();
        this.country = in.readString();
        this.province = in.readString();
        this.city = in.readString();
        this.district = in.readString();
        int tmpWeatherSource = in.readInt();
        this.weatherSource = tmpWeatherSource == -1 ? null : WeatherSource.values()[tmpWeatherSource];
        this.currentPosition = in.readByte() != 0;
        this.residentPosition = in.readByte() != 0;
        this.china = in.readByte() != 0;
    }

    public static final Creator<Location> CREATOR = new Creator<Location>() {
        @Override
        public Location createFromParcel(Parcel source) {
            return new Location(source);
        }

        @Override
        public Location[] newArray(int size) {
            return new Location[size];
        }
    };


}

File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/Location.kt
package wangdaye.com.geometricweather.common.basic.models

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import wangdaye.com.geometricweather.R
import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource
import wangdaye.com.geometricweather.common.basic.models.weather.Astro
import wangdaye.com.geometricweather.common.basic.models.weather.Weather
import java.util.TimeZone
import kotlin.math.abs

class Location(
    val cityId: String,

    val latitude: Float,
    val longitude: Float,
    val timeZone: TimeZone,

    val country: String,
    val province: String? = null,
    val city: String,
    val district: String? = null,

    val weather: Weather? = null,
    val weatherSource: WeatherSource,

    val isCurrentPosition: Boolean = false,
    val isResidentPosition: Boolean = false,
    val isChina: Boolean,
) : Parcelable {

    val formattedId: String
        get() = if (isCurrentPosition) {
            CURRENT_POSITION_ID
        } else {
            cityId + "&" + weatherSource.name
        }

    val isDaylight: Boolean
        get() = isDayLight(this)

    val isUsable: Boolean
        get() = cityId != NULL_ID

    companion object {
        private const val NULL_ID = "NULL_ID"

        const val CURRENT_POSITION_ID = "CURRENT_POSITION"

        @JvmStatic
        @JvmOverloads
        fun isDayLight(location: Location? = null): Boolean {
            val sunRiseProgress = Astro.getRiseProgress(
                astro = location?.weather?.dailyForecast?.getOrNull(0)?.sun,
                timeZone = location?.timeZone ?: TimeZone.getDefault()
            )
            return 0 < sunRiseProgress && sunRiseProgress < 1
        }

        @JvmStatic
        fun buildLocal(): Location {
            return Location(
                cityId = NULL_ID,
                latitude = 0f,
                longitude = 0f,
                timeZone = TimeZone.getDefault(),
                country = "",
                province = "",
                city = "",
                district = "",
                weatherSource = WeatherSource.ACCU,
                isCurrentPosition = true,
                isResidentPosition = false,
                isChina = false
            )
        }

        @JvmStatic
        fun buildDefaultLocation(weatherSource: WeatherSource): Location {
            return Location(
                cityId = "101924",
                latitude = 39.904000f,
                longitude = 116.391000f,
                timeZone = TimeZone.getTimeZone("Asia/Shanghai"),
                country = "中国",
                province = "直辖市",
                city = "北京",
                district = "",
                weatherSource = weatherSource,
                isCurrentPosition = false,
                isResidentPosition = false,
                isChina = true,
            )
        }

        private fun isEquals(a: String?, b: String?): Boolean {
            return if (a.isNullOrEmpty() && b.isNullOrEmpty()) {
                true
            } else if (!a.isNullOrEmpty() && !b.isNullOrEmpty()) {
                a == b
            } else {
                false
            }
        }

        @JvmStatic
        fun excludeInvalidResidentLocation(context: Context, list: List<Location>): List<Location> {
            var currentLocation: Location? = null
            for (l in list) {
                if (l.isCurrentPosition) {
                    currentLocation = l
                    break
                }
            }
            val result = ArrayList<Location>(list.size)
            if (currentLocation == null) {
                result.addAll(list)
            } else {
                for (l in list) {
                    if (!l.isResidentPosition || !l.isCloseTo(context, currentLocation)) {
                        result.add(l)
                    }
                }
            }
            return result
        }

        @JvmStatic
        fun copy(
            src: Location,
            weather: Weather?,
        ) = src.copy(
            weather = weather
        )

        @JvmStatic
        fun copy(
            src: Location,
            weatherSource: WeatherSource,
        ) = src.copy(
            weatherSource = weatherSource
        )

        @JvmStatic
        fun copy(
            src: Location,
            currentPosition: Boolean,
            residentPosition: Boolean,
        ) = src.copy(
            isCurrentPosition = currentPosition,
            isResidentPosition = residentPosition,
        )

        @JvmStatic
        fun copy(
            src: Location,
            latitude: Float,
            longitude: Float,
            timeZone: TimeZone,
        ) = src.copy(
            latitude = latitude,
            longitude = longitude,
            timeZone = timeZone,
        )

        @JvmField
        val CREATOR = object : Parcelable.Creator<Location> {

            override fun createFromParcel(parcel: Parcel): Location {
                return Location(parcel)
            }

            override fun newArray(size: Int): Array<Location?> {
                return arrayOfNulls(size)
            }
        }
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(cityId)
        parcel.writeFloat(latitude)
        parcel.writeFloat(longitude)
        parcel.writeSerializable(timeZone)
        parcel.writeString(country)
        parcel.writeString(province)
        parcel.writeString(city)
        parcel.writeString(district)
        parcel.writeInt(weatherSource.ordinal)
        parcel.writeByte(if (isCurrentPosition) 1 else 0)
        parcel.writeByte(if (isResidentPosition) 1 else 0)
        parcel.writeByte(if (isChina) 1 else 0)
    }

    override fun describeContents(): Int {
        return 0
    }

    constructor(parcel: Parcel) : this(
        cityId = parcel.readString()!!,
        latitude = parcel.readFloat(),
        longitude = parcel.readFloat(),
        timeZone = parcel.readSerializable()!! as TimeZone,
        country = parcel.readString()!!,
        province = parcel.readString(),
        city = parcel.readString()!!,
        district = parcel.readString(),
        weatherSource = WeatherSource.values()[parcel.readInt()],
        isCurrentPosition = parcel.readByte() != 0.toByte(),
        isResidentPosition = parcel.readByte() != 0.toByte(),
        isChina = parcel.readByte() != 0.toByte()
    )

    @JvmOverloads
    fun copy(
        cityId: String? = null,
        latitude: Float? = null,
        longitude: Float? = null,
        timeZone: TimeZone? = null,
        country: String? = null,
        province: String? = null,
        city: String? = null,
        district: String? = null,
        weather: Weather? = null,
        weatherSource: WeatherSource? = null,
        isCurrentPosition: Boolean? = null,
        isResidentPosition: Boolean? = null,
        isChina: Boolean? = null,
    ) = Location(
        cityId = cityId ?: this.cityId,
        latitude = latitude ?: this.latitude,
        longitude = longitude ?: this.longitude,
        timeZone = timeZone ?: this.timeZone,
        country = country ?: this.country,
        province = province ?: this.province,
        city = city ?: this.city,
        district = district ?: this.district,
        weather = weather ?: this.weather,
        weatherSource = weatherSource ?: this.weatherSource,
        isCurrentPosition = isCurrentPosition ?: this.isCurrentPosition,
        isResidentPosition = isResidentPosition ?: this.isResidentPosition,
        isChina = isChina ?: this.isChina
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || other !is Location) {
            return false
        }

        if (formattedId != other.formattedId) {
            return false
        }

        if (isResidentPosition != other.isResidentPosition) {
            return false
        }

        if (weatherSource != other.weatherSource) {
            return false
        }

        val thisWeather = weather
        val otherWeather = other.weather
        if (thisWeather == null && otherWeather == null) {
            return true
        }
        return if (thisWeather != null && otherWeather != null) {
            thisWeather.base.updateDate.time == otherWeather.base.updateDate.time
        } else {
            false
        }
    }

    override fun hashCode(): Int {
        return formattedId.hashCode()
    }

    fun getCityName(context: Context): String {
        val text = if (!district.isNullOrEmpty() && district != "市辖区" && district != "无") {
            district
        } else if (city.isNotEmpty() && city != "市辖区") {
            city
        } else if (!province.isNullOrEmpty()) {
            province
        } else if (isCurrentPosition) {
            context.getString(R.string.current_location)
        } else {
            ""
        }

        if (!text.endsWith(")")) {
            return text
        }
        return text.substringBeforeLast('(')
    }

    override fun toString(): String {
        val builder = StringBuilder("$country $province")
        if (province != city
            && city.isNotEmpty()
        ) {
            builder.append(" ").append(city)
        }
        if (city != district
            && !district.isNullOrEmpty()
        ) {
            builder.append(" ").append(district)
        }
        return builder.toString()
    }

    fun hasGeocodeInformation(): Boolean {
        return (country.isNotEmpty() || !province.isNullOrEmpty()
                || city.isNotEmpty() || !district.isNullOrEmpty())
    }

    private fun isCloseTo(c: Context, location: Location): Boolean {
        if (cityId == location.cityId) {
            return true
        }
        if (isEquals(province, location.province)
            && isEquals(city, location.city)
        ) {
            return true
        }
        return if (isEquals(province, location.province)
            && getCityName(c) == location.getCityName(c)
        ) {
            true
        } else {
            abs(latitude - location.latitude) < 0.8
                    && abs(longitude - location.longitude) < 0.8
        }
    }
}